traversal/expression/Expression.java

```diff
@@ -0,0 +1,199 @@
+            return ValueVertexImpl.Boolean.of(evaluate(vertexMap));
+        }
+
+        @Override
+        default Encoding.ValueType<java.lang.Boolean> returnType() {
+            return Encoding.ValueType.BOOLEAN;
+        }
+
+        @Override
+        default boolean isBoolean() {
+            return true;
+        }
+
+        @Override
+        default Expression.Boolean asBoolean() {
+            return this;
+        }
+    }
+
+    interface Long extends Expression<java.lang.Long> {
+        @Override
+        default ValueVertexImpl.Long evaluateToVertex(Map<Identifier, ValueHolder<?>> vertexMap) {
+            return ValueVertexImpl.Long.of(evaluate(vertexMap));
+        }
+
+        @Override
+        default Encoding.ValueType<java.lang.Long> returnType() {
+            return Encoding.ValueType.LONG;
+        }
+
+        @Override
+        default boolean isLong() {
+            return true;
+        }
+
+        @Override
+        default Expression.Long asLong() {
+            return this;
+        }
+    }
+
+    interface Double extends Expression<java.lang.Double> {
+        @Override
+        default ValueVertexImpl.Double evaluateToVertex(Map<Identifier, ValueHolder<?>> vertexMap) throws TypeDBCheckedException {
+            java.lang.Double result = evaluate(vertexMap);
+            if (result.isInfinite() || result.isNaN()) throw TypeDBCheckedException.of(INVALID_EXPRESSION_RESULT, result);
+            return ValueVertexImpl.Double.of(result);
+        }
+
+        @Override
+        default Encoding.ValueType<java.lang.Double> returnType() {
+            return Encoding.ValueType.DOUBLE;
+        }
+
+        @Override
+        default boolean isDouble() {
+            return true;
+        }
+
+        @Override
+        default Expression.Double asDouble() {
+            return this;
+        }
+    }
+
+    interface String extends Expression<java.lang.String> {
+        @Override
+        default ValueVertexImpl.String evaluateToVertex(Map<Identifier, ValueHolder<?>> vertexMap) throws TypeDBCheckedException {
+            java.lang.String result = evaluate(vertexMap);
+            if (result.length() > SHORT_UNSIGNED_MAX_VALUE) {
+                throw TypeDBCheckedException.of(INVALID_EXPRESSION_RESULT, "String exceeds maximum length of " + SHORT_UNSIGNED_MAX_VALUE);
+            }
+            return ValueVertexImpl.String.of(result);
+        }
+
+        @Override
+        default Encoding.ValueType<java.lang.String> returnType() {
+            return Encoding.ValueType.STRING;
+        }
+
+        @Override
+        default boolean isString() {
+            return true;
+        }
+
+        @Override
+        default Expression.String asString() {
+            return this;
+        }
+    }
+
+    interface DateTime extends Expression<LocalDateTime> {
+        @Override
+        default ValueVertexImpl.DateTime evaluateToVertex(Map<Identifier, ValueHolder<?>> vertexMap) {
```

Comment:
Expressions should not compute Vertices as output directly, they should return values I think

Reply:
I do agree. The actual implementations have signature in `T evaluate(Map<Identifier, ValueHolder<?>> valueMap);`
This is the interfacing for the traversal engine, but that can be pushed out if needed.
I think the traversal engine has always kept vertices as black boxes, so we wouldn't want it to have to deal with the datatypes either.